### PR TITLE
Compile code with Mypyc

### PR DESCRIPTION
**Pull request by ofek**
Hello again! Let me know what you think 😄 

This PR compiles code with [Mypyc](https://github.com/mypyc/mypyc). To do so, I've switched the build system to `hatchling` in order to use the [hatch-mypyc](https://github.com/ofek/hatch-mypyc) build hook plugin. Hatchling is the minimal build backend of [Hatch](https://github.com/ofek/hatch) (v1 rewrite of CLI being released next week). You can think of it as the same distinction between `flit-core` & `flit` or `poetry-core` & `poetry`, with `hatchling` being more similar to `flit-core` in that only accepted standards are used e.g. project metadata format ([PEP 621](https://www.python.org/dev/peps/pep-0621/)), dependency specification ([PEP 631](https://www.python.org/dev/peps/pep-0631/)), etc.

I added binary wheels to the release workflow using the now ubiquitous [cibuildwheel](https://github.com/pypa/cibuildwheel). Here is an example of the files produced: https://pypi.org/project/hatch-showcase/#files

The only thing I couldn't preserve was the hack where we change the `__module__` of `TOMLDecodeError`.

On my Windows machine, `tox -e benchmark` shows a 2-3x performance improvement.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,15 +1,14 @@
...
 
 [project]
 name = "tomli"
-version = "2.0.0"  # DO NOT EDIT THIS LINE MANUALLY. LET bump2version UTILITY DO IT
 description = "A lil' TOML parser"
 authors = [
     { name = "Taneli Hukkinen", email = "hukkin@users.noreply.github.com" },
 ]
-license = { file = "LICENSE" }
+license = "MIT"
```
Comment (ofek) [inline pyproject.toml:11]: Supports https://www.python.org/dev/peps/pep-0639/

File automatically included

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -71,20 +152,14 @@ jobs:
...
+    - uses: actions/download-artifact@v2
       with:
-        python-version: '3.7'
-    - name: Install build and publish tools
-      run: |
-        pip install build twine
-    - name: Build and check
-      run: |
-        rm -rf dist/ && python -m build
-        twine check --strict dist/*
```
Comment (ofek) [inline .github/workflows/tests.yaml:84]: Does this still https://github.com/pypa/gh-action-pypi-publish#disabling-metadata-verification

File: src/tomli/_parser.py [hunk 3]
```diff
@@ -125,10 +125,10 @@ class Flags:
     """Flags that map to parsed keys/namespaces."""
 
     # Marks an immutable namespace (inline array or inline table).
-    FROZEN = 0
+    FROZEN: ClassVar = 0
```
Comment (ofek) [inline src/tomli/_parser.py:128]:
```
tomli\_parser.py:279: error: Cannot access instance attribute "EXPLICIT_NEST" through class object
tomli\_parser.py:279: note: (Hint: Use "x: Final = ..." or "x: ClassVar = ..." to define a class attribute)
tomli\_parser.py:279: error: Cannot access instance attribute "FROZEN" through class object
tomli\_parser.py:281: error: Cannot access instance attribute "EXPLICIT_NEST" through class object
tomli\_parser.py:281: note: (Hint: Use "x: Final = ..." or "x: ClassVar = ..." to define a class attribute)
tomli\_parser.py:297: error: Cannot access instance attribute "FROZEN" through class object
tomli\_parser.py:297: note: (Hint: Use "x: Final = ..." or "x: ClassVar = ..." to define a class attribute)
tomli\_parser.py:302: error: Cannot access instance attribute "EXPLICIT_NEST" through class object
tomli\_parser.py:302: note: (Hint: Use "x: Final = ..." or "x: ClassVar = ..." to define a class attribute)
tomli\_parser.py:320: warning: Treating generator comprehension as list
tomli\_parser.py:323: error: Cannot access instance attribute "EXPLICIT_NEST" through class object
tomli\_parser.py:323: note: (Hint: Use "x: Final = ..." or "x: ClassVar = ..." to define a class attribute)
tomli\_parser.py:327: error: Cannot access instance attribute "EXPLICIT_NEST" through class object
tomli\_parser.py:327: note: (Hint: Use "x: Final = ..." or "x: ClassVar = ..." to define a class attribute)
tomli\_parser.py:329: error: Cannot access instance attribute "FROZEN" through class object
tomli\_parser.py:329: note: (Hint: Use "x: Final = ..." or "x: ClassVar = ..." to define a class attribute)
tomli\_parser.py:342: error: Cannot access instance attribute "FROZEN" through class object
tomli\_parser.py:342: note: (Hint: Use "x: Final = ..." or "x: ClassVar = ..." to define a class attribute)
tomli\_parser.py:437: error: Cannot access instance attribute "FROZEN" through class object
tomli\_parser.py:437: note: (Hint: Use "x: Final = ..." or "x: ClassVar = ..." to define a class attribute)
tomli\_parser.py:453: error: Cannot access instance attribute "FROZEN" through class object
tomli\_parser.py:453: note: (Hint: Use "x: Final = ..." or "x: ClassVar = ..." to define a class attribute)
```

File: src/tomli/_parser.py [hunk 4]
```diff
@@ -174,8 +174,8 @@ def is_(self, key: Key, flag: int) -> bool:
             cont = inner_cont["nested"]
         key_stem = key[-1]
         if key_stem in cont:
-            cont = cont[key_stem]
-            return flag in cont["flags"] or flag in cont["recursive_flags"]
```
Comment (ofek) [inline src/tomli/_parser.py:178]:
```
TypeError: dict object expected; got set
```

File: src/tomli/_parser.py [hunk 5]
```diff
@@ -317,8 +317,8 @@ def key_value_rule(
     key_parent, key_stem = key[:-1], key[-1]
     abs_key_parent = header + key_parent
 
-    relative_path_cont_keys = (header + key[:i] for i in range(1, len(key)))
-    for cont_key in relative_path_cont_keys:
```
Comment (ofek) [inline src/tomli/_parser.py:321]:
```
tomli\_parser.py:320: warning: Treating generator comprehension as list
```

File: src/tomli/__init__.py [hunk 6]
```diff
@@ -4,6 +4,3 @@
 __version__ = "2.0.0"  # DO NOT EDIT THIS LINE MANUALLY. LET bump2version UTILITY DO IT
 
 from ._parser import TOMLDecodeError, load, loads
-
-# Pretend this exception was created here.
-TOMLDecodeError.__module__ = __name__
```
Comment (ofek) [inline src/tomli/__init__.py:9]:
```
tomli\__init__.py:9: error: Only class variables defined as ClassVar can be assigned to
```

**Comment by hukkin** [comment]
Hey, wow, thanks for the PR!

> On my Windows machine, tox -e benchmark shows a 2-3x performance improvement.

Niiice! This should mean almost the same performance as the C++ and Rust extended `pytomlpp` and `rtoml`. With nothing but Python code. Actually pretty amazing!

> Let me know what you think

Unfortunately, I think it may be a bad move to merge this into Tomli right now:
- Tomli has major role in PEP517 based Python bootstrapping right now. Getting that right has involved considering different build-backends, vendoring options, bootstrap processes where PYTHONPATH is manipulated etc. There's many tickets about the struggles we've faced in Tomli's, Flit's and I think even pip's issue tracker. Now that we've got a set-up that seems to satisfy repackagers, I'm hesitant to change anything.
- I believe this would mean a small bit of extra work for repackagers, adapting their processes, which I'd like to spare them from right now.

If https://github.com/hukkin/tomli/issues/141 goes as I hope, it could be very interesting to return to this in a few years, when Tomli's role in the bootstrapping processes is replaced by a stdlib parser.

And yeah thanks for hatch-mypyc in general! Some months ago I was looking for a way to mypycify [mdformat](https://github.com/executablebooks/mdformat) but there wasn't a way that allowed me to keep pyproject.toml, PEP621 metadata etc. I'll probably try these changes out in that repository soon.

**Comment by ofek** [comment]
@hukkin Fixed 😄 I added https://ofek.dev/hatch/latest/config/build/#conditional-execution and now the build hook is disabled by default

**Comment by hukkin** [comment]
Ah sorry for not being clear enough. I think it may not be a good idea to switch build backend right now.

**Comment by ofek** [comment]
Oh I see. Any concern in particular or just not the right time?

**Comment by hukkin** [comment]
My concern is PEP517 bootstrapping issues of distro packagers, where Tomli and its build backend play a big part. The lack of a standard library TOML parser has been a pain for distro packagers, and my understanding is now that flit started vendoring Tomli, they seem somewhat content with the current state. I wouldn't want to disrupt them right now.

Tomli is very widely repackaged now, so changing the build backend means a small bit of extra busywork for a fairly large amount of people.

As soon as a TOML parser moves into the standard library the situation should be very different.

**Comment by ofek** [comment]
Okay, feel free to close this out!

Just so I understand though, can you give an example of what a distro packager would need to do if this was merged? I want `hatchling` to be as easy to use or package as possible. Is it just a matter of adding the ability to e.g. `apt install python-hatchling`?

**Comment by hukkin** [comment]
I'm not a distro packager so I'm not the right person to ask really. But for each package the distros tend to have a file that looks something like this https://src.fedoraproject.org/rpms/python-tomli/blob/rawhide/f/python-tomli.spec where they describe build requirements, how to build, how to install etc. These files may need to adapt if build backend changes. Also for packages like `build`, `flit_core`, `tomli`, also `hatchling` are special cased a lot because they need to be built before the "regular" build tools exist.

> Is it just a matter of adding the ability to e.g. apt install python-hatchling?

Yes this is another thing that needs to happen.

-----
Don't quote me on any of this. I am NOT a distro packager :smile:

**Comment by ofek** [comment]
Thanks! I'll work on that next.

**Pull request closed without merging**